include/miopen/miopen.h

```diff
@@ -4701,6 +4701,106 @@ miopenReduceTensor(miopenHandle_t handle,
 /** @} */
 // CLOSEOUT TensorReduce DOXYGEN GROUP
 
+/**
+ * An miopenProblem will describe a problem for different miopen operations.
+ * For now, this is only used for convolution, but could be used for other
+ * operators in the future(such as GEMM, Pooling, etc)
+ */
+MIOPEN_DECLARE_OBJECT(miopenProblem);
+
+typedef enum
+{
+    miopenProblemDirectionForward        = 0,
+    miopenProblemDirectionBackward       = 1,
+    miopenProblemDirectionBackwardWeight = 2,
+} miopenProblemDirection_t;
+
+// Enums with no meaningful default value should probably have invalid default for fool-proofing.
+typedef enum
+{
+    miopenTensorNameInvalid  = 0,
+    miopenTensorConvolutionX = 1,
+    miopenTensorConvolutionW = 2,
+    miopenTensorConvolutionY = 3,
+} miopenTensorName_t;
+
+typedef enum
+{
+    miopenSearchResultsOrderByTime   = 0,
+    miopenSearchResultsOrderByMemory = 1,
+} miopenSearchResultsOrder_t;
+
+miopenStatus_t miopenCreateConvProblem(miopenProblem_t* problem,
+                                       miopenConvolutionDescriptor_t operatorDesc,
+                                       miopenProblemDirection_t direction);
+
+miopenStatus_t miopenDestroyProblem(miopenProblem_t problem);
+
+// Universal problem should have named tensor fields rather than ordered.
+// Previously, different directions had different input orders, which may lead to confusion.
+// Implicitly ordered fields would also lead to a whole lot of magical ids all over the internals.
```

Comment:
This seems like a comment for internal developers. We should just have a comment for describing this for external users.

Reply:
@pfultz2 Let me argue a bit. Yes, description for external users is necessary, but comments that explain motivation ("rationale") are often very useful for them.